File: OEM/Domain/OperatorPlans/CreatedAt.cs
namespace OEMAPI.Domain.OperatorPlans
{
    public class CreatedAt : ValueObject
    {
        public DateTime Value { get; }

        public CreatedAt(DateTime value)
        {
            if (value == default)
                throw new BusinessRuleValidationException("CreatedAt cannot be default.");

            Value = value;
        }

        protected override IEnumerable<object> GetEqualityComponents()
        {
            yield return Value;
        }
    }
}

File: OEM/src/Domain/OperatorPlans/CreatedAt.ts
import { ValueObject } from "../core/domain/ValueObject";
import { Result } from "../core/logic/Result";
import { Guard } from "../core/logic/Guard";

interface CreatedAtProps {
  value: Date;
}

export class CreatedAt extends ValueObject<CreatedAtProps> {

  get value(): Date {
    return this.props.value;
  }

  private constructor(props: CreatedAtProps) {
    super(props);
  }

  public static create(date: Date): Result<CreatedAt> {
    const guardResult = Guard.againstNullOrUndefined(date, "createdAt");

    if (!guardResult.succeeded) {
      return Result.fail<CreatedAt>(guardResult.message);
    }

    return Result.ok<CreatedAt>(
      new CreatedAt({ value: date })
    );
  }
}
